### DISCUSSION TITLE
On the way of AlpineLinux

### DISCUSSION THREAD

**Discussion by pieroproietti** [pieroproietti/penguins-eggs-legacy#377] [Announcements]
I'm trying to extend `penguins-eggs` to AlpineLinux, something is already working... 

## Alpine colibri 
I build an AlpineLinux following this guide: [WAY-TO-ALPINE](https://github.com/pieroproietti/penguins-eggs/blob/master/DOCS/WAY-TO-ALPINE.md).

## eggs status
![alpine-statujs](https://github.com/user-attachments/assets/78d0ee60-ee92-48a8-82db-462db7d43129)

## boot from live
![biitub-alpine](https://github.com/user-attachments/assets/0bd1b12e-3f10-48c2-973e-348f0e230eb3)

## sidecar.sh
I created an initramfs-lts for live, and put inside `/live/` on the ISO a script named `sidecar.sh`.

After booting the system go in emergency mode. Then, I just create `/mnt` and mount cdrom on it:

```
mkdir /mnt
mount /dev/sr0 /mnt
/mnt/live/sidecar.sh
```
This is the result:

![after-boot](https://github.com/user-attachments/assets/b544f25b-1a0a-431d-8967-b51683d36194)

Unfortunately, while `filestem.squashfs` is correctly mounted RW in `/newroot`, the `switch_root /newroot /sbin/init` is not performed.

I'm writing the progress on [WAY-TO-ALPINE.md](https://github.com/pieroproietti/penguins-eggs/blob/master/DOCS/WAY-TO-ALPINE.md).

# Video
* [Install and building ISO](https://www.youtube.com/watch?v=3MxdBI5fWm8)
* [Testing ISO](https://www.youtube.com/watch?v=3MxdBI5fWm8)

**Comment by pieroproietti** [comment]
The problem seem to be PID, init exit in emergency mode and we lost PID 1

To check the PID, just: ```echo $$```

![pid](https://github.com/user-attachments/assets/3f1f04af-323e-4d79-97d9-179868799514)

At this point we must to start  our `sidecar.sh` inside emergency shell, then go back to init typing 'exit`.

**Reply by pieroproietti** [reply]
I changed again, and rewrote `sidecar.sh` to be used to mount /newroot then   advice to  Type   `exit` to return control to init and continue.


```
#!/bin/sh
#
# At the end of the boot, initram goes into emergency,
# we give the following commands:
# mkdir /mnt
# mount /dev/sr0 /mnt
# /mnt/live/sidecar
#  

# Creating mountpoints
mkdir -p /media/root-ro 
mkdir -p /media/root-rw 

# mount filesystem squashfs on /media/root-ro
mount -t squashfs /mnt/live/filesystem.squashfs /media/root-ro

# Mount tmpfs on /media/root-rw
mount -t tmpfs root-tmpfs /media/root-rw

# Creare i punti di montaggio necessari
mkdir -p /media/root-rw/work 
mkdir -p /media/root-rw/root

# mount  overlayfs on /newroot
mount -t overlay overlay -o lowerdir=/media/root-ro,upperdir=/media/root-rw/root,workdir=/media/root-rw/work /newroot

# mount virtual filesystems on /newroot
mount -t devtmpfs devtmpfs /newroot/dev
mount -t proc proc /newroot/proc
mount -t sysfs sysfs /newroot/sys
mount -t tmpfs tmpfs /newroot/run
cp -r /run/* /newroot/run

echo "sidecar: /newroot was mounted!"
echo "         check it then type 'exit' to continue boot"
```

This is the boot, I see: 
```
sh: can't access to tty : job control turned off
````

After that, I mounted cdrom and started `/mnt/livesidecar.sh`.

![why-dont-fing-init](https://github.com/user-attachments/assets/e4bd0af9-1f29-4f80-81cc-1132cd2b5fbe)

When give `exit` to boot:

```
/sbin/init not found in new root. Launching emergency recovery shell...
```

Of course,  `/sbin/init` exists in `/newroot` and is a link to `/bin/busybox `.

We passed the PID problem, but we get worst! That infamouse destiny... :-(

**Comment by pieroproietti** [comment]
I'm sharing the resulting [experimental ISOs AlpineLinux](https://sourceforge.net/projects/penguins-eggs/files/ISOS/alpine/), to let you to try and look for a solution.

The oldest one mount filesystem on /newroot, actual mount in on /sysroot, I'm unsure on that is the right place.

**Comment by pieroproietti** [comment]
Since the error I get is:
```
/sbin/init not found in new root. Launching emergency recovery shell...
```
I tried replacing `/newroot/sbin/init` with a copy of `/newroot/busybox` but even then it can't find init, which exists!

![cp-busybox](https://github.com/user-attachments/assets/f85236d4-7f6c-4533-945d-7f0a3d34674d)

**Comment by pieroproietti** [comment]
How newroot is mounted:

![mount](https://github.com/user-attachments/assets/d6d1c9a5-6c48-4fcb-acb2-96c0946d38c6)

**Comment by pieroproietti** [comment]
I tried editing `sidecar.sh` by moving the mount point to `/sysroot` where, consulting the code in [mkinitfs](https://github.com/alpinelinux/mkinitfs), it should be correct, and remove mounting of VFS: /proc, /run/, etc. `switch_root must to do that.
```
#!/bin/sh -e

# At the end of the boot, initram goes into emergency,
# we give the following commands:
# mkdir /mnt
# mount /dev/sr0 /mnt
# /mnt/live/sidecar
#  


# Creating mountpoints
mkdir -p /media/root-ro 
mkdir -p /media/root-rw 

# mount filesystem squashfs on /media/root-ro
mount -t squashfs /mnt/live/filesystem.squashfs /media/root-ro

# Mount tmpfs on /media/root-rw
mount -t tmpfs root-tmpfs /media/root-rw

# Creare i punti di montaggio necessari
mkdir -p /media/root-rw/work 
mkdir -p /media/root-rw/root

# mount  overlayfs on /sysroot
mount -t overlay overlay -o lowerdir=/media/root-ro,upperdir=/media/root-rw/root,workdir=/media/root-rw/work /sysroot

# mount virtual filesystems on /sysroot
#mount -t devtmpfs devtmpfs /sysroot/dev
#mount -t proc proc /sysroot/proc
#mount -t sysfs sysfs /sysroot/sys
#mount -t tmpfs tmpfs /sysroot/run
#cp -r /run/* /sysroot/run

echo "sidecar: /sysroot was mounted!"
echo "         check it then type 'exit' to continue boot"
```
Unfortunately, the problem remains exactly the same:

```
/sbin/init not found in new root. Launching emergency recovery shell...
```

Really I don't know where find a solution.

**Comment by pieroproietti** [comment]
To get help I wrote in [Alpine Linux on Reddit](https://www.reddit.com/r/AlpineLinux/comments/1efnzlz/mkinitfs_creating_a_livecd/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)

**Comment by pieroproietti** [comment]
Just had the honor to contact @ncopa the AlpineLinux author, so I think we will found a solution!

**Comment by ncopa** [comment]
I downloaded the ISO and booted it in qemu, and was able to reproduce.

I don't think this can be solved without modifying mkinitfs. The purpose of initramfs is to find and mount the root file system. mkinitfs is designed for the Alpine case, where rootfs is either a block device (eg partition, device-mapper, raid device etc) or a tmpfs system. There is currently no way to tell it to mount a loopback device file within other block device. (eg mount a squashfs file on iso9660 file system. It is simply not designed to do it.

So your options here is to either patch mkinitfs' init, use other initramfs generator that supports this use case, or write your own initramfs init script.

**Reply by pieroproietti** [reply]
I already tried to use dracut on Alpine: seem possible on installed systems it worked without changements, but I was not able to boot from live with it. As far as I remember I used the sidecar.sh way, probably not the right solution.

Fork mkinitfs is easy, franckly put hands inside a bit less.

Let me to think a bit about that way choose.

Thanks a lot

**Comment by pieroproietti** [comment]
I'm restartig to try with dracut, already converted the code to it... until now I can't get module overlayfs installed, we will see.

**Comment by pieroproietti** [comment]
@ncopa @hosseinseilani 

After made an unsuccessfull tempt on Fedora with dracut, I put myself again on Alpine and mkinitfs... and like a young Serendip Prince, I was able to boot from my ISO and revised a little my sidecar.

We must to let init to start his job and try to mount /sysroot - type `exit` on the first emergency shell - then in the new emergency shell start sidecar.sh to mount /sysroot. Then digit `exit` again and the work is done!


There is much to do again, this is - at last - just a workaround... but let me to overpass this step, pratics with Alpine and will see that can happen and if is possible to find collaborations.

If someone want reply the experiment, the ISO is already on [experimental ISOs AlpineLinux](https://sourceforge.net/projects/penguins-eggs/files/ISOS/alpine/), I removed the previous ones.

https://github.com/user-attachments/assets/f0153fd5-4f3c-4294-9511-5bd3fa5487c3

**Comment by pieroproietti** [comment]
Today, for the first time, I was able to remaster and reinstall Alpine Linux using eggs from sources...

Is not complete yet, I must to create an APKBUILD package, but at last is more than I waited in so - relative - short time: about 3 weeks.

I hope this will extend eggs to others distros too, it there is interests.

Video: [remaster and install Linux Alpine](https://youtu.be/zjev4Zg9sHM)

**Comment by pieroproietti** [comment]
Package penguins-eggs for Alpine is not so bad - but for me is difficult and need time or help - so this morning I just created a manual installation of penguins-eggs, to let curious to remaster and install their Alpine derivatives distros.

eggs at last is a node program and all end in `/usr/lib/penguins-eggs`, that we need are the package node, pnpm and various packages used by eggs. It'is all on [WAY-TO-ALPINE](https://github.com/pieroproietti/penguins-eggs/blob/master/DOCS/WAY-TO-ALPINE.md).

So I changed a bit the install-eggs-dev to create this structure and link to it. This ha the vantage it's on /usr so it's present on the live too.

Now when we start our remastered ISO, we can simply install the system using krill, the eggs CLI installer.

There is again same way to do, but we can already look for new horizonts. A new [video](https://youtu.be/t44vLVmxBlI)

**Comment by pieroproietti** [comment]
I did the Alpine package for penguins-eggs on my fork and  I'm waiting will be accepted on master. This is the  [status](https://gitlab.alpinelinux.org/pieroproietti/aports/-/commit/2087e9a35239c315c48da7dcee4c79e806759357), read all discussion, I explain the problems I had.

In the while is possible install penguins-eggs on Alpine usig my [get-eggs](https://github.com/pieroproietti/get-eggs).

Package released for Alpine is version 10.0.28, but in the while I'm working on calamares too for Alpine.

I was not able to get calamares running on Alpine 3.20, so I decide to try with the compiled version on 3.18.

Something is happening... the program load, the configurations, is almost OK: eggs now can install/remove calamares and configure it... 

![alpine318-calamares](https://github.com/user-attachments/assets/880ae009-1add-4597-b3bc-ddc39a20e508)

Note: calamares stop, unpacking the filesystem.

You can test it downloading the colibri-318 edition on my [sf page](https://sourceforge.net/projects/penguins-eggs/files/ISOS/alpine/).

**Reply by pieroproietti** [reply]
![calamares-error](https://github.com/user-attachments/assets/d112a481-0ac6-4f6f-bbb8-34a5eb7f489d)

This is the error I get:
```
<div><strong>&lt;class 'Boost.Python.ArgumentError'&gt;</strong></div><div>Python argument types in
    libcalamares.utils.mount(str, str, str, list)
did not match C++ signature:
    mount(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; device_path, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; mount_point)
    mount(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; device_path, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; mount_point, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; filesystem_name)
    mount(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; device_path, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; mount_point, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; filesystem_name, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; options)</div><div><br/>Traceback:</div><div><pre>File &quot;/usr/lib/calamares/modules/mount/main.py&quot;, line 251, in run
    mount_partition(root_mount_point, partition, partitions)

  File &quot;/usr/lib/calamares/modules/mount/main.py&quot;, line 179, in mount_partition
    if libcalamares.utils.mount(device,
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</pre></div>
       ```

**Comment by pieroproietti** [comment]
Same suggestions? I feel me a bit alone in this adventure... It's OK, but...
